fist/serializer.c

```diff
@@ -29,29 +46,79 @@ void sdump(hashmap *hmap) {
                 keyval object = on.maps[key];
                 uint32_t length = object.key.length;
                 // Writes key length and key name to db file
-                fwrite(&length, sizeof(length), 1, dump);
-                fwrite(dtext(object.key), object.key.length, 1, dump);
+
+                size += fwrite(&length, sizeof(length), 1, dump);
+                size += fwrite(dtext(object.key), object.key.length, 1, dump);
 
                 uint32_t num_values = object.values.length;
                 // Writes number of values associated with key to db file
-                fwrite(&num_values, sizeof(num_values), 1, dump);
+                size += fwrite(&num_values, sizeof(num_values), 1, dump);
                 for(int value = 0; value < object.values.length; value++) {
                     // Writes value to db file
                     dstring value_on = object.values.values[value];
                     uint32_t val_length = value_on.length;
-                    fwrite(&val_length, sizeof(val_length), 1, dump);
-                    fwrite(dtext(value_on), value_on.length, 1, dump);
+
+                    size += fwrite(&val_length, sizeof(val_length), 1, dump);
+                    size += fwrite(dtext(value_on), value_on.length, 1, dump);
                 }
             }
         }
     }
+
+    fseek(dump, 0, SEEK_END);
+    uint64_t len = ftell(dump);
+    fseek(dump, 0, SEEK_SET);
+    unsigned char *buffer = malloc(len);
+    fread(buffer, 1, len, dump);
+
+    sdump_compress(buffer, len);
+
     fclose(dump);
 }
 
+FILE *sload_compressed() {
+    FILE *db;
+    if((db = fopen("fist.db", "rb"))) {
+        uint64_t original_size;
+        fread(&original_size, 8, 1, db);
+
+        fseek(db, 0, SEEK_END);
+        long length = ftell(db);
+        fseek(
+            db, 0,
+            SEEK_SET); // Skip first 8 bytes which are to indicate the original size of the db file.
+
+        length -= 8;
+        fseek(db, 8, SEEK_CUR);
+
+        unsigned char *data = malloc(length);
+        fread(data, 1, length, db);
+
+        unsigned char *decompressed = malloc(original_size);
+        int size;
+
+        if(!(size = lzf_decompress(data, length, decompressed, original_size))) {
+            printf("Error decompressing DB file");
+            return 0;
+        }
+
+        fclose(db);
+
+        FILE *tmp = tmpfile();
+        fwrite(decompressed, original_size, 1, tmp);
+        rewind(tmp);
```

Comment:
Instead of rewinding and reusing the old code, would it be better to just work directly on the decompressed buffer?

Reply:
Being able to read the the data using `fread()` makes it much easier. On most systems my understanding is that tmpfiles are created in memory. Other than on the systems that do not do this, would there be a significant speedup in using just the `decompressed` buffer and rewriting the deserialization function?